typed-racket-lib/typed-racket/typecheck/tc-toplevel.rkt

```diff
@@ -315,11 +315,20 @@
               (~datum prefix-all-defined) (~datum prefix-all-defined-except)
               (~datum expand)))))
 
+;; Move type declarations to the beginning of the list,
+;;  keep other declarations in the same relative order.
+;; (-> (Listof Syntax) (Listof Syntax))
+(define (lift-type-declarations form*)
+  (define (is-type-decl? form)
+    (syntax-parse form [_:type-declaration #t] [_ #f]))
+  (let*-values ([(type-decl* form*) (partition is-type-decl? form*)])
+    (append type-decl* form*)))
+
 ;; actually do the work on a module
 ;; produces prelude and post-lude syntax objects
 ;; syntax-list -> (values syntax syntax)
 (define (type-check forms0)
-  (define forms (syntax->list forms0))
+  (define forms (lift-type-declarations (syntax->list forms0)))
```

Comment:
Possible issue: I'm re-ordering the top-level forms so that type declarations come first. Without this, type annotations aren't associated with the right definitions.

Not sure whether this will change to fix the bug with structs.